### PR TITLE
User-defined annotations

### PR DESCRIPTION
**Pull request by asterite**
This PR adds user-defined **annotations** to the language.

Annotations are the **old attributes**:

```crystal
@[Link("sqlite3")] # <- this is an annotation 
lib LibSQLite3
end
```

I chose to use the name "annotation" instead of "attribute" because the latter is much more common in code: I searched usage of "annotation" and it was all over the place in Crystal projects, whereas I could just find one project that used the name "annotation" somewhere. So to avoid breaking a lot of projects, "annotation" might be a better name.

It works like this.

First you define an annotation (just like any other type):

```crystal
module JSON
  # Put this on top of instance variables to control serialization
  anntotation Field
  end
end
```

Note that `annotation` now becomes a keyword, just like `class` and `def`, so it's prohibited mostly everywhere, so that's why I chose to go with that name, because it's pretty rare.

Once you define an annotation, you can use it:

```crystal
class Point1D
  @[JSON::Field(key: "ex")] 
  @x : Int32
end
```

And then you can query it:

```crystal
class Point1D
  def do_something
    {% for ivar in @type.instance_vars %}
      {% ann = ivar.annotation(::JSON::Field) %}
      {% key = ((ann && ann[:key]) || ivar).id %}
      # do something with key
    {% end %}
  end
end
```

In the above example it looks like JSON serialization becomes a lot more tedious than the current `JSON.mapping`. However, the way I imagine it, you would include a `JSON::Serializable` module, which would define the `initialize` and `to_json` methods, and by default will use all instance variables, even if they don't have the `JSON::Field` annotation on them. The annotation is only used to **control** the serialization, like choosing the `key`, or whether to ignore a field.

You can compile the compiler with this PR and try it out, [here's a working prototype](https://gist.github.com/asterite/db0ee7947d7eceb17c894fabffffafc4).

Annotations can also be placed on top of types, and later queried. Methods for now can't have (user-defined) annotations, but maybe in the future they will.

The benefits of having annotations is that it provides a way to attach metadata to instance variables. Right now one has to use a compile-time constant, add info to that, and later process it in a `finished` macro, which isn't ideal (though depending on what you want to do, because of [this chicken and egg problem](https://github.com/crystal-lang/crystal/issues/6028), you might need to still do it like that).

Other benefits, which I will highlight just by using the `JSON::Serialization` implementation using annotations:
- it works well with inheritance and included modules
- you can make a `record` include `JSON::Serializable` and not have to repeat the `JSON.mapping` declaration

Of course this will also apply to YAML serialization and any other formats. And I'm sure you will come up with great ways to use annotations.

Now, the annotation **declaration** is empty right now:

```crystal
annotation Foo
  # Nothing can go in here
end
```

In the future, we could specify what things an annotation can be applied to, or what are the fields it can hold. But since it's mainly a compile-time thing that you'll interact with in macros, and macros are pretty lax (they feel like a dynamic Crystal), it makes sense not to restrict that. That's why you can still specify a converter like `@[JSON::Field(converter: Time::Format.new("%F %T"))]`, where it's not clear what type `converter` has (like this, anything is basically an ASTNode, which is what macros work with).

A few other things:
- this PR fixes a bug related to getting the default value of an instance variable at compile-time (related to inheritance and modules)
- this PR adds the `TypeNode#nilable?` macro method, which I think is very useful (and used in my prototype json serializable annotation)
- the prototype has a hack somewhere: I tried to fix it but I couldn't understand why it failed. So for now, we can use that workaround, maybe later with time and patience someone can fix it
- someone will have to update the current docs regarding all of this ;-)

If we merge this, just in the next release we'll be able to define the `JSON::Serializable` module and all that stuff. We _could_ do it now by using `compare_version` and defining stuff conditionally based on the next compiler version, but I'd rather wait for a release to be able to try and review stuff more easily.

See also: #3620, #3325, #5643

### CODE REVIEW

File: src/compiler/crystal/types.cr [hunk 1]
```diff
@@ -162,6 +162,10 @@ module Crystal
       self.is_a?(NilType)
     end
 
+    def nilable?
+      self.is_a?(NilType) || (self.is_a?(UnionType) && self.union_types.any?(&.nil_type?))
```
[the remark this answers is not in the document]
Comment (asterite) [reply src/compiler/crystal/types.cr]: A union type can only have single types, it doesn't recurse (or it doesn't need to recurse).

**Comment by RX14** [comment]
I'm definitely not liking the mismatch of naming between annotations definition and their usage in code. If it takes a massive hack in the compiler (hardcode `Primitive` until next release) to do it without having to suffix annotations with `Annotation`, it's definitely worth it to me.

File: src/compiler/crystal/types.cr (hunk 1 above)
[the remark this answers is not in the document]
Comment (bew) [reply src/compiler/crystal/types.cr]: Oh ok, didn't know that

**Comment by asterite** [comment]
@RX14 As I explain somewhere above, I was actually going to do it that way (with the `Annotation` suffix) to avoid cluttering the top-level namespace. Then I said "Meh, let's leave that for later" and then I immediately found a conflict, which only means that avoiding that name clash is actually something good.

Plus, again, it's how it's done in C# (only in C# you can use the full attribute name too, when using it, but I don't think we need that in Crystal, or at least we could maybe easily support it later).

But I wouldn't mind not using the suffix and waiting a bit. It's just that seeing `Link` there in the API docs, openning and seeing "well, it's a link *attribute*, not a link class or struct or module" is probably a bit confusing. It's much clear if it's called `LinkAttribute`. The example I gave about `XML::Element` and `XML::ElementAttribute` is algo a good one. And in my `JSON::Serializable` prototype there is a `JSON::SerializableAttribute` too, to specify `strict`ness of parsing, and it just looks nice and consistent.

Well, that's my argument anyway :-)

**Comment by oprypin** [comment]
I agree with RX14.

I don't know about conflicts, maybe it's good to prevent having a class and an annotation under the same name?

Or you could just make a hack that appends `Annotation` to all names, both in `annotation` and `@[ ]` (the latter is actually already done!)

**Comment by oprypin** [comment]
Your argument about discerning annotations in docs is not an argument towards particular naming, it's really just a feature request for docs generator.

**Comment by asterite** [comment]
That said, if everyone else also dislikes this "hack", for now, I can hardcode that "Primitive" always refer to the primitive annotation, when used in annotations, and later remove that rule.

**Comment by kostya** [comment]
thanks, this json serialization, i dream about:
```crystal
class JsonWithDefaults
  include JSON::Serializable

  property a = 11
  property b = "Haha"
  property c = true
  property d = false
  property e : Bool? = false
  property f : Int32? = 1
  property g : Int32?
  property h = [1, 2, 3]
end
```

if we also can generate initialize with named_tuple it would be super.

**Comment by asterite** [comment]
@kostya that's possible. What do you mean?

**Comment by kostya** [comment]
```crystal
class JsonWithDefaults
  include Initialize
  include JSON::Serializable

  property a : Int32
  property b = "Haha"
end

JsonWithDefaults.new(a: 5, b: "bla")
JsonWithDefaults.new(a: 5)
JsonWithDefaults.from_json(%<{"a":5}>)
```

**Comment by asterite** [comment]
Please add an emoji in this comment for:
- 👍 to **keep** the requirement of the "Annotation" suffix
- 👎 to **remove** the requirement of the "Annotation" suffix

But if we go with 👎 , we'll have to think of another name of the attribute used on top of a type to let the parsing be `strict` (though now that I think about it, there was this idea of having a method you could override to handle unknown attributes, which by default would just `pull.skip`... so I guess this is also a 👍 for 👎 from my part 😛).

**Comment by asterite** [comment]
I just pushed a commit that removes the requirement that annotation names must end with "Annotation". I had to hardcode two things in annotations, `Primitive` and `Flags`, because there's `Int::Primitive` and `File::Info::Flags`, and there isn't yet a way to write `::Primitive` and `::Flags` in annotations, but in the next version we can use that and remove the hardcoded rule.

File: src/compiler/crystal/semantic/semantic_visitor.cr [hunk 2]
```diff
@@ -419,18 +419,27 @@ abstract class Crystal::SemanticVisitor < Crystal::Visitor
...
-    path = ann.path.clone
-    path.names[-1] = "#{path.names.last}Annotation"
-
-    type = lookup_type(path)
+    # Since there's `Int::Primitive`, and now we'll have
+    # `::Primitive`, but there's no way to specify ::Primitive
+    # just yet in annotations, we temporarily hardcode
+    # that `Primitive` inside annotations means the top
+    # level primitive.
+    # We also have the same problem with File::Flags, which
```
Comment (RX14) [inline src/compiler/crystal/semantic/semantic_visitor.cr]: Could you put a `TODO:` string in here so we can grep for all todos and not miss this?

Comment (asterite) [reply src/compiler/crystal/semantic/semantic_visitor.cr]: Done!

File: src/enum.cr [hunk 3]
```diff
@@ -26,7 +26,7 @@
 #
 # ### Flags enum
 #
-# An enum can be marked with the `@[Flags]` attribute. This changes the default values:
+# An enum can be marked with the `FlagsAnnotation` annotation. This changes the default values:
```
Comment (r00ster91) [inline src/enum.cr]: I think `FlagsAnnotation` should be changed to `Flags` here.

**Comment by asterite** [comment]
They are in the same namespace, yes. Maybe later we can put them in a separate namespace, I don't know. But then, if they are in a different namespace, how would you show them in docs? That's a bit ugly.

That's why I first suggested with stick with the "Annotation" suffix, but nobody liked that. But I don't think it's a big deal, really.

**Comment by bew** [comment]
We could put them in a special `Crystal::Annotation` namespace, and the doc generator would just have to check in this namespace

**Comment by asterite** [comment]
But annotations can appear under other namespaces, like in the `JSON::Field` example above. And you can define top-level annotations if you want to.

**Comment by asterite** [comment]
@kostya Hey, I found a way to do what you want with the **current** (0.24.2) version of Crystal 😄 

```crystal
class Object
  module Initialize
    def initialize(**args : **T) forall T
      {% for key in T.keys.map(&.id) %}
        {% unless @type.instance_vars.map(&.id).includes?(key) %}
          {% raise "no argument named '#{key}'" %}
        {% end %}

        @{{key}} = args[{{key.symbolize}}]
      {% end %}
    end
  end
end

class Foo
  include Initialize

  property id : Int32
  property x = 1
  property y = "foo"
end

p Foo.new(id: 1)
p Foo.new(id: 2, y: "hi")
p Foo.new(id: 2, x: 3)
p Foo.new(id: 3, x: 10, y: "hello")
```

And you even get a "nice" error message if you use it in a wrong way:

```
Error in bar.cr:27: no argument named 'z'

p Foo.new(z: 10)
      ^~~
```

**Comment by bew** [comment]
@asterite true, what about keeping the `Annotation` suffix internally (but keep `annotation Field` & `@[Field]` clean)

**Comment by asterite** [comment]
Then it will clash with stuff that ends with Annotation.

Let's leave it like this, there's no problem, really.

**Comment by kostya** [comment]
strange bug: https://play.crystal-lang.org/#/r/40ab,

**Comment by straight-shoota** [comment]
IMHO the behaviour of such edge cases should at least be defined.

I'd expect using the same annotation type multiple times to raise an error instead of simply using the last one. That's for example how Java does it.

And class variables should either not allow annotations or accept them but they can't be used, that would be nice to be in a spec.

**Comment by asterite** [comment]
> I'd expect using the same annotation type multiple times to raise an error instead of simply using the last one. That's for example how Java does it.

Java also doesn't let you reopen classes or redefine methods.

> And class variables should either not allow annotations or accept them but they can't be used, that would be nice to be in a spec.

A class variable can be marked as `ThreadLocal`, that's the only attribute allowed for now. I guess we can raise an error if it's some other attribute. And do the same for methods (which can accept a bunch of predefined attributes).

File: spec/compiler/codegen/macro_spec.cr [hunk 4]
```diff
@@ -1749,6 +1749,27 @@ describe "Code gen: macro" do
       )).to_i.should eq(10)
   end
 
+  it "gets default value of instance variable of inherited type that also includes module" do
```
[the remark this answers is not in the document]
Comment (asterite) [reply spec/compiler/codegen/macro_spec.cr]: I'm not sure. This is about the default value (that was the bug), and "default value" doesn't appear your suggested sentence.

File: src/compiler/crystal/semantic/top_level_visitor.cr [hunk 5]
```diff
@@ -181,6 +155,41 @@ class Crystal::TopLevelVisitor < Crystal::SemanticVisitor
     scope.types[name] = type
     node.resolved_type = type
 
+    process_annotations(annotations) do |annotation_type, ann|
+      if node.struct? && type.is_a?(NonGenericClassType)
+        case annotation_type
+        when @program.extern
```
[the remark this answers is not in the document]
Comment (asterite) [reply src/compiler/crystal/semantic/top_level_visitor.cr]: Good idea. I'll change them all, but I think `@program.extern_annotation`might be more clear.

**Comment by asterite** [comment]
@straight-shoota I added a few more commits to address the issues you pointed.

File: spec/compiler/codegen/macro_spec.cr (hunk 4 above)
[the remark this answers is not in the document]
Comment (straight-shoota) [reply spec/compiler/codegen/macro_spec.cr]: Ah, okay. I figured the default value was only used as a vehicle to test `instance_vars`.

File: src/compiler/crystal/codegen/link.cr [hunk 6]
```diff
@@ -172,15 +167,15 @@ module Crystal
...
     end
 
-    private def add_link_attributes(types, attrs)
+    private def add_link_annotations(types, annotations)
       types.try &.each_value do |type|
         next if type.is_a?(AliasType) || type.is_a?(TypeDefType)
 
-        if type.is_a?(LibType) && type.used? && (link_attrs = type.link_attributes)
-          attrs.concat link_attrs
+        if type.is_a?(LibType) && type.used? && (links_annotations = type.link_annotations)
```
Comment (straight-shoota) [inline src/compiler/crystal/codegen/link.cr]: `links_annotations` doesn't need a doubled plural

**Comment by straight-shoota** [comment]
> Java also doesn't let you reopen classes or redefine methods.

Yeah, that's true. I wasn't thinking about reopening or redefining but if the same annotation is specified at the same location multiple times:
```cr
@[Foo(1)]
@[Foo(2)]
class Bar
  @[Foo(1)]
  @[Foo(2)]
  @bar : Bool
end
```

That's where I'd raise an error because it clearly makes no sense.

**Comment by asterite** [comment]
@straight-shoota That's also the case for:

```crystal
def foo(x)
end

# This should be an error, this is pointless
def foo(x)
end
```

but since it's not an error (though I think maybe it should be), I'd say we should fix all of those cases in a separate PR. But note that this is a bit hard, sometimes these things might get defined and redefined via macros. So erring on the permissive side is better in my opinion.

**Comment by asterite** [comment]
@RX14 Bump! :-)

**Comment by RX14** [comment]
Looks good!

Couldn't we add an `annotation` macro method on `Def` to get annotations to cover methods at least?

**Comment by straight-shoota** [comment]
I think `annotation Foo; end` is better for the moment, it leave open the option to add something. But that can be evaluated separately. If we decide to never put something inside a annotation definition, it's easy to remove `; end`.

If we strip it right now and later decide to add a body, code will break without a nice error message pointing to the problem.

**Comment by asterite** [comment]
I just added a commit to allow annotations on methods, and being able to query them.

Yes, my reasoning with `annotation Foo; end` is exactly that. If it ends up like that, well, it's just a little more typing. But if not, we can easily extend that syntax with other stuff.

**Comment by RX14** [comment]
I can imagine exactly what data we want to put in an annotation def - my fear is just that we never get around to it :)

**Comment by asterite** [comment]
Things to specify can be:
- Where the annotation can be used (instance var, method, etc.)
- What's the shape of the annotation arguments (for example one could specify the named argument names at least, so you'll get a fast and nice error message when used incorrectly)

**Comment by RX14** [comment]
Let's get a new JSON interface based on this PR merged, then we can cut 0.25 I think.

**Comment by Sija** [comment]
@RX14 Based on the sheer amount of features it should be more like 0.30 🍘

**Comment by bew** [comment]
@RX14 iirc you can't directly use the compiler feature until next release, so the new JSON interface will have to wait next release

**Comment by bew** [comment]
But maybe we can release a 0.24.3 with annotations (only?), so that we can develop that new serialization interface for 0.25 ?

**Comment by asterite** [comment]
This can be done. If someone wants to work on this, let us know.

Here's a diff to do it:

```diff
diff --git a/src/json.cr b/src/json.cr
index ce859e2d7..58abfd904 100644
--- a/src/json.cr
+++ b/src/json.cr
@@ -93,3 +93,7 @@ module JSON
 end
 
 require "./json/*"
+
+{% if compare_versions(Crystal::VERSION, "0.25.0") >= 0 %}
+  require "./json/next/serialization"
+{% end %}
diff --git a/src/json/next/serialization.cr b/src/json/next/serialization.cr
new file mode 100644
index 000000000..0b14ba494
--- /dev/null
+++ b/src/json/next/serialization.cr
@@ -0,0 +1,7 @@
+module JSON
+  annotation Field
+  end
+
+  module Serializable
+  end
+end
```

In short:
1. We conditionally require a file based on the crystal version
2. We use the new features in this file

I put it in `json/next/serialization.cr` because `json.cr` does `require "./json/*"` so we must put it in a nested directory.

For specs it'll be a bit trickier.

Alternatively, there's no reason why we can't release 0.25.0 "soon", and then work on 0.26.0 next and release it soon too. I used to released maybe once a month in the past. We can release as often as we'd like (except that I'm not in charge of releasing anymore).

**Comment by kostya** [comment]
i add some initial work on it: https://github.com/crystal-lang/crystal/pull/6082

**Comment by RX14** [comment]
Release is easier but it still needs someone from manas. @bcardiff should be around in a week or so to cut a release...

**Comment by RX14** [comment]
Maybe we should start a release PR now though, to track anything left we want in `0.25.0` and write the changelog.

**Comment by asterite** [comment]
Can we ask them to give you, RX14, or someone else from the core team, to do releases? At this point this is the only thing that's holding us back from managing Crystal as a true open source community project.

**Comment by RX14** [comment]
@asterite yeah I think thats a good idea. I'd really like crystal to be disconnected from manas in terms of management. Perhaps I'll ask @bcardiff to walk me through next release?

**Pull request merged**